File: backtracking_nqueen.cpp
#include <iostream>
#include <vector>

using namespace std;

#define N 4 // 4x4 Chessboard

// Function to print the chessboard solution
void printSolution(vector<vector<int>>& board) {
    for (int i = 0; i < N; i++) {
        cout << "  ";
        for (int j = 0; j < N; j++) {
            if (board[i][j] == 1) cout << "Q ";
            else cout << ". ";
        }
        cout << endl;
    }
    cout << "-----------------------" << endl;
}

// Function to check if a queen can be placed on board[row][col]
bool isSafe(vector<vector<int>>& board, int row, int col) {
    int i, j;

    // Check this row on left side
    for (i = 0; i < col; i++)
        if (board[row][i] == 1) return false;

    // Check upper diagonal on left side
    for (i = row, j = col; i >= 0 && j >= 0; i--, j--)
        if (board[i][j] == 1) return false;

    // Check lower diagonal on left side
    for (i = row, j = col; j >= 0 && i < N; i++, j--)
        if (board[i][j] == 1) return false;

    return true;
}

// Recursive utility function to solve N-Queen
bool solveNQUtil(vector<vector<int>>& board, int col) {
    // Base Case: If all queens are placed
    if (col >= N) return true;

    // Consider this column and try placing this queen in all rows one by one
    for (int i = 0; i < N; i++) {
        if (isSafe(board, i, col)) {
            // Place this queen
            board[i][col] = 1;

            // Recur to place rest of the queens
            if (solveNQUtil(board, col + 1)) return true;

            // BACKTRACK: If placing queen in board[i][col] doesn't lead to a solution
            board[i][col] = 0;
        }
    }

    return false;
}

int main() {
    cout << "--- DSA Masterclass: N-Queen Problem (C++) ---" << endl << endl;

    vector<vector<int>> board(N, vector<int>(N, 0));

    if (!solveNQUtil(board, 0)) {
        cout << "Solution does not exist" << endl;
    } else {
        cout << "One of the valid configurations for 4-Queens:" << endl;
        printSolution(board);
    }

    return 0;
}

File: backtracking_nqueen.py
def print_solution(board):
    for row in board:
        print("  " + " ".join(["Q" if x == 1 else "." for x in row]))
    print("-----------------------")

def is_safe(board, row, col, n):
    # Check this row on left side
    for i in range(col):
        if board[row][i] == 1:
            return False

    # Check upper diagonal on left side
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False

    # Check lower diagonal on left side
    for i, j in zip(range(row, n, 1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False

    return True

def solve_nq_util(board, col, n):
    # Base Case: All queens are placed
    if col >= n:
        return True

    for i in range(n):
        if is_safe(board, i, col, n):
            # Place queen
            board[i][col] = 1

            # Recur to place rest
            if solve_nq_util(board, col + 1, n):
                return True

            # BACKTRACK
            board[i][col] = 0

    return False

def main():
    print("--- DSA Masterclass: N-Queen Problem (Python) ---\n")
    
    n = 4
    board = [[0] * n for _ in range(n)]

    if not solve_nq_util(board, 0, n):
        print("Solution does not exist")
    else:
        print("One of the valid configurations for 4-Queens (Python):")
        print_solution(board)

if __name__ == "__main__":
    main()